core/src/retries.rs

```diff
@@ -485,4 +485,68 @@ mod tests {
         listener_task.abort();
         assert!(listener_task.await.unwrap_err().is_cancelled());
     }
+
+    #[tokio::test]
+    async fn http_retry_server_json_problem() {
+        install_test_trace_subscriber();
+        let mut server = mockito::Server::new_async().await;
+
+        let mock_problem = server
+            .mock("GET", "/")
+            .with_status(400)
+            .with_header("Content-Type", "application/problem+json")
+            .with_body("{\"type\":\"evil://league_of_evil/bad.horse?hes.bad\"}")
+            .expect(1)
+            .create_async()
+            .await;
+
+        let http_client = reqwest::Client::builder().build().unwrap();
+
+        let response = retry_http_request(LimitedRetryer::new(0), || async {
+            http_client.get(server.url()).send().await
+        })
+        .await
+        .unwrap_err()
+        .unwrap();
+
+        assert_eq!(response.status(), StatusCode::BAD_REQUEST);
+        assert_eq!(
+            response.type_uri(),
+            Some("evil://league_of_evil/bad.horse?hes.bad")
+        );
+        mock_problem.assert_async().await;
+    }
+
+    #[tokio::test]
+    async fn http_retry_server_json_problem_versioned_doc() {
+        install_test_trace_subscriber();
+        let mut server = mockito::Server::new_async().await;
+
+        // Ensure that even with a complex media type, we parse a problem
+        // document and prefer its status code.
```

Comment:
This part of the comment is incorrect, we happen to prefer the HTTP status code over the `status` member of the details. The latter is mostly advisory, see https://www.rfc-editor.org/rfc/rfc9457.html#name-status.
```suggestion
        // document.
```